File: frontend/src/lib/api/abp.ts
import { api } from './client'
import { ABPJugada, ABPRivalJugada, ABPPartidoJugada, ABPSesionJugada, ABPPartidoPlanFull } from '@/types'

export interface ABPJugadaCreateData {
  equipo_id?: string
  nombre: string
  codigo?: string
  tipo: string
  lado: string
  subtipo?: string
  descripcion?: string
  senal_codigo?: string
  sistema_marcaje?: string
  notas_tacticas?: string
  fases?: any[]
  asignaciones?: any[]
  es_plantilla?: boolean
  tags?: string[]
  orden?: number
}

export interface ABPJugadaUpdateData extends Partial<ABPJugadaCreateData> {}

export interface ABPRivalJugadaCreateData {
  nombre: string
  tipo: string
  lado: string
  subtipo?: string
  descripcion?: string
  fases?: any[]
  video_url?: string
  tags?: string[]
}

export const abpApi = {
  // ============ Biblioteca CRUD ============

  async list(equipo_id: string, params?: { tipo?: string; lado?: string; subtipo?: string; busqueda?: string }): Promise<{ data: ABPJugada[] }> {
    return api.get<{ data: ABPJugada[] }>('/abp', {
      params: { equipo_id, ...params },
    })
  },

  async get(id: string): Promise<ABPJugada> {
    return api.get<ABPJugada>(`/abp/${id}`)
  },

  async create(data: ABPJugadaCreateData): Promise<ABPJugada> {
    return api.post<ABPJugada>('/abp', data)
  },

  async update(id: string, data: ABPJugadaUpdateData): Promise<ABPJugada> {
    return api.put<ABPJugada>(`/abp/${id}`, data)
  },

  async delete(id: string): Promise<void> {
    return api.delete(`/abp/${id}`)
  },

  async duplicate(id: string): Promise<ABPJugada> {
    return api.post<ABPJugada>(`/abp/${id}/duplicate`, {})
  },

  // ============ Partido ============

  async listPartido(partidoId: string): Promise<{ data: ABPPartidoJugada[] }> {
    return api.get<{ data: ABPPartidoJugada[] }>(`/abp/partido/${partidoId}`)
  },

  async assignToPartido(partidoId: string, data: { jugada_id: string; notas?: string; orden?: number; asignaciones_override?: any[] }): Promise<ABPPartidoJugada> {
    return api.post<ABPPartidoJugada>(`/abp/partido/${partidoId}`, data)
  },

  async unassignFromPartido(partidoId: string, jugadaId: string): Promise<void> {
    return api.delete(`/abp/partido/${partidoId}/${jugadaId}`)
  },

  // ============ Partido Plan (workflow) ============

  async getPlan(partidoId: string): Promise<ABPPartidoPlanFull> {
    return api.get<ABPPartidoPlanFull>(`/abp/partido/${partidoId}/plan`)
  },

  async savePlan(partidoId: string, data: {
    comentario_ofensivo?: string
    comentario_defensivo?: string
    jugadas: { jugada_id: string; asignaciones_override?: any[]; notas?: string; orden: number }[]
  }): Promise<ABPPartidoPlanFull> {
    return api.put<ABPPartidoPlanFull>(`/abp/partido/${partidoId}/plan`, data)
  },

  // ============ Rival ============

  async listRival(rivalId: string, params?: { tipo?: string; lado?: string }): Promise<{ data: ABPRivalJugada[] }> {
    return api.get<{ data: ABPRivalJugada[] }>(`/abp/rival/${rivalId}`, { params })
  },

  async createRival(rivalId: string, data: ABPRivalJugadaCreateData): Promise<ABPRivalJugada> {
    return api.post<ABPRivalJugada>(`/abp/rival/${rivalId}`, data)
  },

  async updateRival(rivalId: string, jugadaId: string, data: Partial<ABPRivalJugadaCreateData>): Promise<ABPRivalJugada> {
    return api.put<ABPRivalJugada>(`/abp/rival/${rivalId}/${jugadaId}`, data)
  },

  async deleteRival(rivalId: string, jugadaId: string): Promise<void> {
    return api.delete(`/abp/rival/${rivalId}/${jugadaId}`)
  },

  // ============ Sesion ============

  async listSesion(sesionId: string): Promise<{ data: ABPSesionJugada[] }> {
    return api.get<{ data: ABPSesionJugada[] }>(`/abp/sesion/${sesionId}`)
  },

  async linkToSesion(sesionId: string, data: { jugada_id: string; notas?: string; orden?: number }): Promise<ABPSesionJugada> {
    return api.post<ABPSesionJugada>(`/abp/sesion/${sesionId}`, data)
  },

  async unlinkFromSesion(sesionId: string, jugadaId: string): Promise<void> {
    return api.delete(`/abp/sesion/${sesionId}/${jugadaId}`)
  },

  // ============ PDF ============

  async downloadPlaybookPdf(equipo_id: string, params?: { tipo?: string; lado?: string }): Promise<Blob> {
    return api.getBlob('/abp/playbook-pdf', { params: { equipo_id, ...params } })
  },

  async downloadPartidoPdf(partidoId: string): Promise<Blob> {
    return api.getBlob(`/abp/partido/${partidoId}/pdf`)
  },
}

File: backend/app/models/abp.py
"""
TrainingHub Pro - Modelos ABP (Acciones a Balón Parado)
Schemas Pydantic para jugadas de balón parado.
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime
from uuid import UUID
from enum import Enum


class TipoABP(str, Enum):
    CORNER = "corner"
    SEMI_CORNER = "semi_corner"
    FALTA_LATERAL = "falta_lateral"
    FALTA_FRONTAL = "falta_frontal"
    FALTA_LEJANA = "falta_lejana"
    PENALTI = "penalti"
    SAQUE_BANDA = "saque_banda"
    SAQUE_PUERTA = "saque_puerta"


class LadoABP(str, Enum):
    OFENSIVO = "ofensivo"
    DEFENSIVO = "defensivo"


class SubtipoABP(str, Enum):
    INSWING = "inswing"
    OUTSWING = "outswing"
    CORTO = "corto"
    DIRECTO = "directo"
    INDIRECTO = "indirecto"
    LARGO = "largo"


class SistemaMarcaje(str, Enum):
    ZONAL = "zonal"
    INDIVIDUAL = "individual"
    MIXTO = "mixto"


class ABPPlayerRol(str, Enum):
    LANZADOR = "lanzador"
    BLOQUEADOR = "bloqueador"
    PALO_CORTO = "palo_corto"
    PALO_LARGO = "palo_largo"
    BORDE_AREA = "borde_area"
    SENUELO = "señuelo"
    RECHACE = "rechace"
    REFERENCIA = "referencia"
    BARRERA = "barrera"
    MARCAJE_ZONAL = "marcaje_zonal"
    MARCAJE_INDIVIDUAL = "marcaje_individual"
    PORTERO = "portero"
    OTRO = "otro"


# ============ Schemas de Fase ============

class ABPFase(BaseModel):
    id: str
    nombre: str = Field(..., max_length=100)
    orden: int = 0
    descripcion: Optional[str] = None
    diagram: dict = Field(default_factory=lambda: {
        "elements": [], "arrows": [], "zones": [], "pitchType": "half"
    })


class ABPAsignacion(BaseModel):
    element_id: str
    jugador_id: Optional[str] = None
    rol: Optional[str] = None


# ============ Schemas de Jugada (Biblioteca) ============

class ABPJugadaBase(BaseModel):
    nombre: str = Field(..., min_length=1, max_length=255)
    codigo: Optional[str] = Field(None, max_length=20)
    tipo: str
    lado: str
    subtipo: Optional[str] = None
    descripcion: Optional[str] = None
    senal_codigo: Optional[str] = Field(None, max_length=50)
    sistema_marcaje: Optional[str] = None
    notas_tacticas: Optional[str] = None
    fases: List[ABPFase] = Field(default_factory=list)
    asignaciones: List[ABPAsignacion] = Field(default_factory=list)
    es_plantilla: bool = False
    tags: List[str] = Field(default_factory=list)
    orden: int = 0


class ABPJugadaCreate(ABPJugadaBase):
    equipo_id: Optional[str] = None


class ABPJugadaUpdate(BaseModel):
    nombre: Optional[str] = Field(None, min_length=1, max_length=255)
    codigo: Optional[str] = Field(None, max_length=20)
    tipo: Optional[str] = None
    lado: Optional[str] = None
    subtipo: Optional[str] = None
    descripcion: Optional[str] = None
    senal_codigo: Optional[str] = Field(None, max_length=50)
    sistema_marcaje: Optional[str] = None
    notas_tacticas: Optional[str] = None
    fases: Optional[List[ABPFase]] = None
    asignaciones: Optional[List[ABPAsignacion]] = None
    es_plantilla: Optional[bool] = None
    tags: Optional[List[str]] = None
    orden: Optional[int] = None


class ABPJugadaResponse(ABPJugadaBase):
    id: UUID
    organizacion_id: UUID
    equipo_id: Optional[UUID] = None
    creado_por: Optional[UUID] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ============ Schemas de Rival Jugada ============

class ABPRivalJugadaBase(BaseModel):
    nombre: str = Field(..., min_length=1, max_length=255)
    tipo: str
    lado: str
    subtipo: Optional[str] = None
    descripcion: Optional[str] = None
    fases: List[ABPFase] = Field(default_factory=list)
    video_url: Optional[str] = None
    tags: List[str] = Field(default_factory=list)


class ABPRivalJugadaCreate(ABPRivalJugadaBase):
    pass


class ABPRivalJugadaUpdate(BaseModel):
    nombre: Optional[str] = Field(None, min_length=1, max_length=255)
    tipo: Optional[str] = None
    lado: Optional[str] = None
    subtipo: Optional[str] = None
    descripcion: Optional[str] = None
    fases: Optional[List[ABPFase]] = None
    video_url: Optional[str] = None
    tags: Optional[List[str]] = None


class ABPRivalJugadaResponse(ABPRivalJugadaBase):
    id: UUID
    rival_id: UUID
    organizacion_id: UUID
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ============ Schemas de Partido-Jugada ============

class ABPPartidoJugadaCreate(BaseModel):
    jugada_id: str
    asignaciones_override: Optional[List[ABPAsignacion]] = None
    notas: Optional[str] = None
    orden: int = 0


class ABPPartidoJugadaResponse(BaseModel):
    id: UUID
    partido_id: UUID
    jugada_id: UUID
    asignaciones_override: Optional[List[Any]] = None
    notas: Optional[str] = None
    orden: int = 0
    created_at: Optional[datetime] = None
    jugada: Optional[ABPJugadaResponse] = None

    class Config:
        from_attributes = True


# ============ Schemas de Sesion-Jugada ============

class ABPSesionJugadaCreate(BaseModel):
    jugada_id: str
    notas: Optional[str] = None
    orden: int = 0


class ABPSesionJugadaResponse(BaseModel):
    id: UUID
    sesion_id: UUID
    jugada_id: UUID
    notas: Optional[str] = None
    orden: int = 0
    created_at: Optional[datetime] = None
    jugada: Optional[ABPJugadaResponse] = None

    class Config:
        from_attributes = True


# ============ Schemas de Partido Plan ============

class ABPPartidoPlanJugadaItem(BaseModel):
    jugada_id: str
    asignaciones_override: Optional[List[ABPAsignacion]] = None
    notas: Optional[str] = None
    orden: int = 0


class ABPPartidoPlanFullSave(BaseModel):
    comentario_ofensivo: Optional[str] = None
    comentario_defensivo: Optional[str] = None
    jugadas: List[ABPPartidoPlanJugadaItem] = Field(default_factory=list)


class ABPPartidoPlanResponse(BaseModel):
    id: Optional[UUID] = None
    partido_id: UUID
    comentario_ofensivo: Optional[str] = None
    comentario_defensivo: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ABPPartidoPlanFullResponse(BaseModel):
    plan: Optional[ABPPartidoPlanResponse] = None
    jugadas: List[ABPPartidoJugadaResponse] = Field(default_factory=list)
